File: src/server/events.ts
// Server-sent events for the desktop window: after every capture, new or existing, the
// window moves to the item's reader page (desktop/src-tauri follows `open-reader`).
import type { Context } from "hono";
import { streamSSE } from "hono/streaming";

export type OpenReaderEvent = { reader_url: string };

// Bun closes a connection that stays idle for 10 s; a comment line between events keeps
// the stream open so no event falls into a reconnect gap.
const KEEPALIVE_INTERVAL_MS = 5_000;

export class BucketEvents {
  private readonly subscribers = new Set<(event: OpenReaderEvent) => void>();

  publishOpenReader(event: OpenReaderEvent): void {
    for (const send of this.subscribers) {
      send(event);
    }
  }

  stream(c: Context): Response {
    return streamSSE(c, async (stream) => {
      const send = (event: OpenReaderEvent) => {
        void stream.writeSSE({ event: "open-reader", data: JSON.stringify(event) });
      };
      this.subscribers.add(send);
      stream.onAbort(() => {
        this.subscribers.delete(send);
      });
      while (!stream.aborted) {
        await stream.sleep(KEEPALIVE_INTERVAL_MS);
        await stream.write(": keepalive\n\n");
      }
    });
  }
}

File: server/src/events.rs
//! Server-sent events for the library: after every capture, new or existing, each open library
//! opens the item's reader in a tab (src/web/tabs.tsx), and the desktop window comes to the
//! front (desktop/src-tauri follows `open-reader`).
use std::convert::Infallible;

use axum::extract::State;
use axum::response::sse::{Event, KeepAlive, Sse};
use futures::stream::Stream;
use futures::StreamExt;
use tokio::sync::broadcast;
use tokio_stream::wrappers::BroadcastStream;

use crate::config::EVENT_KEEPALIVE;
use crate::contract::OpenReader;
use crate::state::Shared;

pub struct Events {
    open_reader: broadcast::Sender<OpenReader>,
}

impl Events {
    pub fn new() -> Self {
        let (open_reader, _) = broadcast::channel(16);
        Self { open_reader }
    }

    /// Tells every open library to show the reader.
    pub fn publish_open_reader(&self, event: OpenReader) {
        match self.open_reader.send(event) {
            Ok(_windows) => {}
            // No window is listening: the capture stands, and the next window opens the library.
            Err(broadcast::error::SendError(_unheard)) => {}
        }
    }
}

impl Default for Events {
    fn default() -> Self {
        Self::new()
    }
}

pub async fn stream(
    State(state): State<Shared>,
) -> Sse<impl Stream<Item = Result<Event, Infallible>>> {
    let events = BroadcastStream::new(state.events.open_reader.subscribe()).filter_map(
        |received| async move {
            // A subscriber that fell behind the channel skips what it missed.
            let event = match received {
                Ok(event) => event,
                Err(tokio_stream::wrappers::errors::BroadcastStreamRecvError::Lagged(_)) => {
                    return None
                }
            };
            let data = serde_json::to_string(&event).expect("an event serializes");
            Some(Ok(Event::default().event("open-reader").data(data)))
        },
    );
    Sse::new(events).keep_alive(KeepAlive::new().interval(EVENT_KEEPALIVE).text("keepalive"))
}
